tao_compiler/mlir/disc/transforms/disc_shape_optimization.cc

```diff
@@ -301,13 +316,181 @@ struct ExtractElementOfTieShapeOpCanonicalizationPattern
+    if (!indexOp) return failure();
+    int64_t index = indexOp.getValue().cast<IntegerAttr>().getInt();
+
+    auto concatOp = op.tensor().getDefiningOp<mhlo::ConcatenateOp>();
+    if (!concatOp) return failure();
+    if (!isCandidateShapeTensorType(op.tensor().getType())) return failure();
+
+    for (auto& en : llvm::enumerate(concatOp->getOperands())) {
+      if (!isCandidateShapeTensorType(en.value().getType())) return failure();
+      auto operandTy = en.value().getType().cast<RankedTensorType>();
+      if (index >= operandTy.getNumElements()) {
+        index -= operandTy.getNumElements();
+        continue;
+      }
+
+      Value newIndex =
+          rewriter.create<arith::ConstantIndexOp>(op.getLoc(), index);
+      Value newValue =
+          rewriter.create<tensor::ExtractOp>(op.getLoc(), en.value(), newIndex);
+      rewriter.replaceOp(op, {newValue});
+      return success();
+    }
+    return failure();
+  }
+};
+
+// convert:
+//   %1 = mhlo.slice(%0) {start_indices = dense<1>, ...}
+//         : (tensor<3xi32>) -> tensor<2xi32>
+//   %2 = tensor.extract %1[%c1] : ...
+// to:
+//   %2 = tensor.extract %0[%c2] : ...
+struct ExtractElementOfSliceOpCanonicalizationPattern
+    : public OpRewritePattern<tensor::ExtractOp> {
+  using OpRewritePattern<tensor::ExtractOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tensor::ExtractOp op,
+                                PatternRewriter& rewriter) const override {
+    if (op.indices().size() != 1) return failure();
+    auto indexOp = dyn_cast_or_null<arith::ConstantOp>(
+        op.indices().front().getDefiningOp());
+    if (!indexOp) return failure();
+    int64_t index = indexOp.getValue().cast<IntegerAttr>().getInt();
+
+    auto sliceOp = op.tensor().getDefiningOp<mhlo::SliceOp>();
+    if (!sliceOp) return failure();
+    if (!isCandidateShapeTensorType(op.tensor().getType())) return failure();
+
+    int64_t start = sliceOp.start_indices().getValues<int64_t>()[0];
+    int64_t stride = sliceOp.strides().getValues<int64_t>()[0];
+    int64_t inputIndex = (start + index * stride);
+    Value newIndex =
+        rewriter.create<arith::ConstantIndexOp>(op.getLoc(), inputIndex);
+    Value newValue = rewriter.create<tensor::ExtractOp>(
+        op.getLoc(), sliceOp.operand(), newIndex);
+
+    rewriter.replaceOp(op, {newValue});
+    return success();
+  }
+};
+
+// convert:
+//   %1 = mhlo.reshape(%0) : (tensor<i32>) -> tensor<1xi32>
+//   %2 = tensor.extract %1[%c0] : ...
+// to:
+//   %2 = tensor.extract %0[] : ...
+struct ExtractElementOfReshapeOpCanonicalizationPattern
+    : public OpRewritePattern<tensor::ExtractOp> {
+  using OpRewritePattern<tensor::ExtractOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tensor::ExtractOp op,
+                                PatternRewriter& rewriter) const override {
+    auto tensorTy = op.tensor().getType().dyn_cast<RankedTensorType>();
+    if (!tensorTy || !tensorTy.hasStaticShape() ||
```

Comment:
why it's different from `isCandidateShapeTensorType`?

Reply:
Reshape is more special in our use cases. For example:

```
%0 = reshape from tensor<i32> to tensor<1xi32>
%1 = reshape from tensor<i32> to tensor<1xi32>
%2 = mhlo.concat(%0, %1) : tensor<2xi32>
```
Here the input of reshape has rank zero, which is usually not considered as a shape tensor. Thus, we need to relax the restrictions of reshape here.